Forge/src/main/java/me/ferdz/placeableitems/client/renderer/tileentity/FluidHolderRenderer.java

```diff
@@ -0,0 +1,176 @@
+ *
+ * @author Parker Hawke - Choco
+ */
+public class FluidHolderRenderer extends TileEntityRendererFast<FluidHolderTileEntity> {
+
+    private final Map<PlaceableItemsBlock, FluidModel> fluidModels = new IdentityHashMap<>();
+
+    @Override
+    public void renderTileEntityFast(FluidHolderTileEntity tile, double x, double y, double z, float partialTicks, int destroyStage, BufferBuilder buffer) {
+        FluidStack fluidStack = tile.getFluid();
+        if (fluidStack.isEmpty()) {
+            return;
+        }
+
+        BlockState state = tile.getBlockState();
+        Block block = state.getBlock();
+        if (!(block instanceof PlaceableItemsBlock)) {
+            return;
+        }
+
+        FluidHolderBlockComponent fluidComponent = getFluidHolderBlockComponent((PlaceableItemsBlock) block);
+        Preconditions.checkState(fluidComponent != null, "FluidHolderTileEntity does not have a FluidHolderBlockComponent. This is impossible.");
+        if (!fluidComponent.shouldRenderFluid()) {
+            return;
+        }
+
+        FluidModel model = this.fluidModels.get(block);
+        Preconditions.checkState(model != null, "Missing model binding for FluidHolderTileEntity (" + tile.getClass().getName() + ")");
+
+        Vector3d rotationPoint = model.getRotationPoint(state);
+        buffer.setTranslation(x + (rotationPoint.x / 16F), y + (rotationPoint.y / 16F), z + (rotationPoint.z / 16F));
+
+        // Render the fluid
+        Fluid fluid = fluidStack.getFluid();
+        TextureAtlasSprite still = Minecraft.getInstance().getTextureMap().getAtlasSprite(fluid.getAttributes().getStill(fluidStack).toString());
+        float u1 = still.getMinU(), u2 = still.getMaxU();
+        float v1 = still.getMinV(), v2 = still.getMaxV();
+
+        int color = fluid.getAttributes().getColor(fluidStack);
+        float alpha = (color >> 24 & 0xFF) / 255F;
+        float red = (color >> 16 & 0xFF) / 255F;
+        float green = (color >> 8 & 0xFF) / 255F;
+        float blue = (color & 0xFF) / 255F;
+
+        BlockPos pos = tile.getPos();
+        AllVertexBoundingBox bounds = model.getRenderBounds(state);
+
+        if (model.shouldRender(state, Direction.DOWN)) {
+            int downCombined = getWorld().getCombinedLight(pos.down(), 0);
+            int downLMa = downCombined >> 16 & 65535;
+            int downLMb = downCombined & 65535;
+
+            bufferPos(buffer, bounds.getBackBottomLeft()).color(red, green, blue, alpha).tex(u1, v2).lightmap(downLMa, downLMb).endVertex();
+            bufferPos(buffer, bounds.getFrontBottomLeft()).color(red, green, blue, alpha).tex(u1, v1).lightmap(downLMa, downLMb).endVertex();
+            bufferPos(buffer, bounds.getFrontBottomRight()).color(red, green, blue, alpha).tex(u2, v1).lightmap(downLMa, downLMb).endVertex();
+            bufferPos(buffer, bounds.getBackBottomRight()).color(red, green, blue, alpha).tex(u2, v2).lightmap(downLMa, downLMb).endVertex();
+        }
```

Comment:
Would you mind adding a little bit of comments explaining how this works? I'm not familiar with the whole rendering part

Reply:
Absolutely. It's quad rendering so it does seem more confusing than it is, but I'll comment what I can